feat: add support for internal temperature sensor (tsens) for esp32c6 and esp32c3

davoclavo opened the pull request:
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [ ] I have added necessary changes to user code to the [Migration Guide](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/MIGRATING-0.21.md).
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/documentation/CONTRIBUTING.md) and followed its instructions.

### Pull Request Details 📖

#### Description
Added basic support for internal temperature sensor for esp32c6 and esp32c3. Starts paving the path to fix https://github.com/esp-rs/esp-hal/issues/2116

Next steps:
- Add temperature ranges to the peripheral config
- Add interrupt handlers
- Support more chipsets

#### Testing
I created an example that measures internal chip temperature every second and prints it.

Any feedback is more than welcome, I'm very happy to start contributing to this amazing project!

davoclavo commented on esp-hal/src/tsens.rs:
```diff
@@ -0,0 +1,117 @@
+
+        // We need to enable ApbSarAdc clock before trying to write on the tsens_ctrl
+        // register
+        PeripheralClockControl::enable(crate::system::Peripheral::ApbSarAdc);
```
What is the best way to handle this requirement of enabling `Peripheral::AbpSarAdc` before modifying the power up register for `tsens_ctrl`?

MabezDev replied:
I think we should also guard this with a `GenericPeripheralGuard`, as iirc the ADC uses this, and we probably don't want to reset & enable if it's already in use.

AnthonyGrondin commented on esp-hal/src/tsens.rs:
```diff
@@ -0,0 +1,146 @@
+
+        let measurement = abp_saradc.tsens_ctrl().read().out().bits();
+
+        // TODO Address multiple temperature ranges and offsets
```
Thank you for your contribution,

I don't know if that's still the case, but last year, there was an effort to remove all float usage from `esp-rs` because it increases the binary size by around 2k after optimization.

See: https://github.com/esp-rs/esp-hal/issues/108

davoclavo replied:
Thanks a lot for the heads up. 

Should I just return the raw `u8` and document the formula needed to convert it to a usable temperature? or is there a better approach for this issue, such as hiding these functionalities under a cargo feature?

bjoernQ replied:
I could imagine offering something like `raw_temperature` and documenting the formula as you said.

Additonally, you could offer `celcius`, `fahrenheit` etc. which does fixed-point math for convenience - or probably even better we get a new `Temperature` type doing fixed-point math and offering conversions

In any case we shouldn't add a feature for this

MabezDev replied:
We an keep the float API, as long as we provide the raw value too. If a user doesnt want to pull in floats they can use the raw API, and in theory (toolchains willing of course) the float machinery won't be brought in.

MabezDev commented on esp-hal/src/tsens.rs:
```diff
@@ -0,0 +1,146 @@
+/// Temperature sensor configuration
+#[derive(Debug, Clone, Default, PartialEq, Eq, Copy, procmacros::BuilderLite)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct Config {
```
This should be marked as `#[non_exaustive]`

MabezDev replied:
and add a hash derive please

MabezDev commented on esp-hal/src/tsens.rs:
```diff
@@ -0,0 +1,146 @@
+
+/// Temperature sensor configuration error
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ConfigError {}
```
This should be marked as `#[non_exaustive]`

MabezDev replied:
and add a `Hash` derive please

MabezDev commented on esp-hal/src/tsens.rs:
```diff
@@ -0,0 +1,146 @@
+pub enum ConfigError {}
+
+/// Temperature sensor driver
+pub struct TemperatureSensor<'d> {
```
Missing a `Debug` derive here.

MabezDev commented on examples/src/bin/temperature_sensor.rs:
```diff
@@ -0,0 +1,33 @@
+//! This example uses the internal temperature sensor to measure the chip's temperature
```
Can you move this to `qa-test` instead? There isn't much benefit to this as an example (over what the docs provide), but adding a temp sens to HIL test might be tricky... or useful to detecting a fire in the server room :D

MabezDev commented:
Thanks for the PR! Overall this a nice PR, and pretty close already - just a few bits to iron out.

davoclavo commented:
Thanks a lot for the feedback. I have addressed all the suggested changes/improvements mentioned and rebased with the lastest changes.

About adding a HIL test: I tried to add a deterministic test to ensure the sensor is properly configured and working. Thought about making a test that tries to heat up the CPU and test there was an increase in temperature, but couldn't reliably heat it up in my experiments.

bugadani commented on esp-hal/src/lib.rs:
```diff
@@ -194,6 +194,9 @@ pub mod debugger;
 pub mod time;
 
+#[cfg(tsens)]
+pub mod tsens;
```
This needs to be in the unstable module list.

davoclavo replied:
Done 👍🏼

davoclavo commented:
Thanks a lot for the feedback! will keep it in mind, and looking forward to keep contributing :)
